x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java

```diff
@@ -0,0 +1,119 @@
+        ROOT_RESOURCE_PATH + "meta_index_template.json", Version.CURRENT.toString(), VERSION_PATTERN,
+        Collections.singletonMap(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id)));
+
+    private static final IndexTemplateConfig NOTIFICATIONS_TEMPLATE = new IndexTemplateConfig(NotificationsIndex.NOTIFICATIONS_INDEX,
+        ROOT_RESOURCE_PATH + "notifications_index_template.json", Version.CURRENT.toString(), VERSION_PATTERN,
+        Collections.singletonMap(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id)));
+
+    private static final IndexTemplateConfig CONFIG_TEMPLATE = configTemplate();
+
+    private static final IndexTemplateConfig INFERENCE_TEMPLATE = new IndexTemplateConfig(InferenceIndexConstants.LATEST_INDEX_NAME,
+        ROOT_RESOURCE_PATH + "inference_index_template.json", Version.CURRENT.toString(), VERSION_PATTERN,
+        Collections.singletonMap(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id)));
+
+    private static IndexTemplateConfig configTemplate() {
+        Map<String, String> variables = new HashMap<>();
+        variables.put(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id));
+        variables.put(Pattern.quote("${xpack.ml.config.max_result_window}"),
+            String.valueOf(AnomalyDetectorsIndex.CONFIG_INDEX_MAX_RESULTS_WINDOW));
+        variables.put(Pattern.quote("${xpack.ml.config.mappings}"), MlConfigIndex.mapping());
+
+        return new IndexTemplateConfig(AnomalyDetectorsIndex.configIndexName(),
+            ROOT_RESOURCE_PATH + "config_index_template.json",
+            Version.CURRENT.toString(), VERSION_PATTERN,
+            variables);
+    }
+
+    private static IndexTemplateConfig anomalyDetectionResultsTemplate() {
+        Map<String, String> variables = new HashMap<>();
+        variables.put(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id));
+        variables.put(Pattern.quote("${xpack.ml.anomalydetection.results.mappings}"), AnomalyDetectorsIndex.resultsMapping());
+
+        return new IndexTemplateConfig(AnomalyDetectorsIndex.jobResultsIndexPrefix(),
+            ANOMALY_DETECTION_PATH + "results_index_template.json",
+            Version.CURRENT.toString(), VERSION_PATTERN,
+            variables);
+    }
+
+    private final boolean mlEnabled;
+
+    public MlIndexTemplateRegistry(Settings nodeSettings, ClusterService clusterService, ThreadPool threadPool, Client client,
+                                   NamedXContentRegistry xContentRegistry) {
+        super(nodeSettings, clusterService, threadPool, client, xContentRegistry);
+        this.mlEnabled = MachineLearning.ML_ENABLED.get(nodeSettings);
+    }
+
+    @Override
+    protected List<IndexTemplateConfig> getTemplateConfigs() {
+        if (mlEnabled == false) {
+            return Collections.emptyList();
```

Comment:
This is a difference compared to what we did with the old `getIndexTemplateMetaDataUpgrader()` functionality.  In that we used to return our templates even when ML was disabled on the node.  Returning nothing when ML is disabled isn't necessarily wrong, but it's a difference that needs thinking about.

There may be some subtle effect in Cloud for example.  I think at one time there was a proposal that when a Cloud cluster gets switched to different hardware and a full cluster snapshot from the old hardware is restored into brand new empty cluster running on different hardware that all the X-Pack plugins are disabled during that restore.  I am not sure if this is still being proposed or not.  But it's a scenario where the ML plugin would be disabled in a cluster that contained ML indices.

Let's consult more widely on this offline.

Reply:
It was wrong to return an empty list. A number of tests was failing because of that. As long as the plugin is loaded, the templates should be there as some nodes may become ML nodes.